catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/ResultIterable.java

```diff
@@ -0,0 +1,165 @@
+ * </p>
+ */
+public class ResultIterable implements Iterable<Result> {
+
+    private final QueryFunction queryFunction;
+
+    private final QueryRequest queryRequest;
+
+    /**
+     * Creates an iterable that will call the {@link CatalogFramework} to retrieve the results
+     * that match the {@link QueryRequest} provided.
+     *
+     * @param catalogFramework reference to the {@link CatalogFramework} to call to retrieve the
+     *                         results.
+     * @param queryRequest     request used to retrieve the results.
+     */
+    public ResultIterable(CatalogFramework catalogFramework, QueryRequest queryRequest) {
+        notNull(catalogFramework, "Catalog framework reference cannot be null");
+        notNull(queryRequest, "Query request cannot be null");
+
+        this.queryFunction = catalogFramework::query;
+        this.queryRequest = queryRequest;
+    }
+
+    /**
+     * Creates an iterable that will call a {@link QueryFunction} to retrieve the results
+     * that match the {@link QueryRequest} provided.
+     *
+     * @param queryFunction reference to the {@link QueryFunction} to call to retrieve the
+     *                      results.
+     * @param queryRequest  request used to retrieve the results.
+     */
+    public ResultIterable(QueryFunction queryFunction, QueryRequest queryRequest) {
+        notNull(queryFunction, "Query function cannot be null");
+        notNull(queryRequest, "Query request cannot be null");
+
+        this.queryFunction = queryFunction;
+        this.queryRequest = queryRequest;
+    }
+
+    @Override
+    public Iterator<Result> iterator() {
+        return new QueryResultIterator(queryFunction, queryRequest);
+    }
+
+    private static class QueryResultIterator implements Iterator<Result> {
+
+        private final QueryFunction queryFunction;
+
+        private int currentIndex;
+
+        private QueryImpl queryCopy;
+
+        private QueryRequestImpl queryRequestCopy;
+
+        private Iterator<Result> results = Collections.emptyIterator();
+
+        QueryResultIterator(QueryFunction queryFunction, QueryRequest queryRequest) {
+            this.queryFunction = queryFunction;
+
+            copyQueryRequestAndQuery(queryRequest);
+
+            this.currentIndex = queryCopy.getStartIndex();
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (results.hasNext()) {
+                return true;
+            }
+
+            fetchNextResults();
+
+            return results.hasNext();
+        }
+
+        @Override
+        public Result next() {
+            if (results.hasNext()) {
+                return results.next();
+            }
+
+            fetchNextResults();
+
+            if (!results.hasNext()) {
+                throw new NoSuchElementException("No more results match the specified query");
+            }
+
+            return results.next();
+        }
+
+        private void fetchNextResults() {
```

Comment:
If it is a federated query, does the consumer have to use the method `CatalogFramework. query(QueryRequest fedQueryRequest)`? That does not match the signature of the functional interface because of the second parameter.